kernel/constr.ml

```diff
@@ -435,6 +435,24 @@ let fold f acc c = match kind c with
   | CoFix (_,(lna,tl,bl)) ->
     Array.fold_left2 (fun acc t b -> f (f acc t) b) acc tl bl
 
+let fold_with_binders g f n acc c = match kind c with
+  | (Rel _ | Meta _ | Var _   | Sort _ | Const _ | Ind _
+    | Construct _) -> acc
+  | Cast (c,_,t) -> f n (f n acc c) t
+  | Prod (_,t,c) -> f (g n) (f n acc t) c
+  | Lambda (_,t,c) -> f (g n) (f n acc t) c
+  | LetIn (_,b,t,c) -> f (g n) (f n (f n acc b) t) c
+  | App (c,l) -> Array.fold_left (f n) (f n acc c) l
+  | Proj (p,c) -> f n acc c
+  | Evar (_,l) -> Array.fold_left (f n) acc l
+  | Case (_,p,c,bl) -> Array.fold_left (f n) (f n (f n acc p) c) bl
```

Comment:
I don't think this is the natural order of subterms: the user sees `c` before `p`.
This glitch is not only in this code, but since you are adding in a place were it will be found and used, I think it should be made right.

Reply:
> This glitch is not only in this code, but since you are adding in a place were it will be found and used, I think it should be made right.

OK.

By the way, when the comment in constr.mli says 
```
[fold f acc c] folds [f] on the immediate subterms of [c]
   starting from [acc] and proceeding from left to right according to
   the usual representation of the constructions
```
does it mean the user syntax or the ocaml constructor?